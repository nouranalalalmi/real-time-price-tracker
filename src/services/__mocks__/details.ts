export const useGetAssetDetails = jest.fn(() => ({
  data: {
    id: 'bitcoin',
    rank: '1',
    symbol: 'BTC',
    name: 'Bitcoin',
    supply: '19693006.0000000000000000',
    maxSupply: '21000000.0000000000000000',
    marketCapUsd: '1154471549507.9040036854136828',
    volumeUsd24Hr: '13866421449.4822673574551073',
    priceUsd: '58623.4295316775917138',
    changePercent24Hr: '2.2406824871311752',
    vwap24Hr: '57798.6225887850445657',
    explorer: 'https://blockchain.info/',
  },
  isLoading: false,
}));

export const useGetAssetHistory = jest.fn(() => ({
  data: [
    {
      priceUsd: '57453.4654235888369902',
      time: 1714575600000,
      circulatingSupply: '19692650.0000000000000000',
      date: '2024-05-01T15:00:00.000Z',
    },
    {
      priceUsd: '56849.8205946134484453',
      time: 1714579200000,
      circulatingSupply: '19692650.0000000000000000',
      date: '2024-05-01T16:00:00.000Z',
    },
    {
      priceUsd: '56989.2744984085723538',
      time: 1714582800000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T17:00:00.000Z',
    },
    {
      priceUsd: '57421.7196556879817478',
      time: 1714586400000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T18:00:00.000Z',
    },
    {
      priceUsd: '58915.9403189355241213',
      time: 1714590000000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T19:00:00.000Z',
    },
    {
      priceUsd: '57038.2007750086044715',
      time: 1714593600000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T20:00:00.000Z',
    },
    {
      priceUsd: '57342.4529782201663178',
      time: 1714597200000,
      circulatingSupply: '19692737.0000000000000000',
      date: '2024-05-01T21:00:00.000Z',
    },
    {
      priceUsd: '57871.1674520177401757',
      time: 1714600800000,
      circulatingSupply: '19692737.0000000000000000',
      date: '2024-05-01T22:00:00.000Z',
    },
    {
      priceUsd: '57904.7652767478518852',
      time: 1714604400000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-01T23:00:00.000Z',
    },
    {
      priceUsd: '58350.5875656442168466',
      time: 1714608000000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T00:00:00.000Z',
    },
    {
      priceUsd: '57759.0173401927487826',
      time: 1714611600000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T01:00:00.000Z',
    },
    {
      priceUsd: '57173.5786733097720706',
      time: 1714615200000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T02:00:00.000Z',
    },
    {
      priceUsd: '57451.8361411585783057',
      time: 1714618800000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T03:00:00.000Z',
    },
    {
      priceUsd: '57495.4104137510215865',
      time: 1714622400000,
      circulatingSupply: '19692837.0000000000000000',
      date: '2024-05-02T04:00:00.000Z',
    },
    {
      priceUsd: '57362.3002691941893159',
      time: 1714626000000,
      circulatingSupply: '19692837.0000000000000000',
      date: '2024-05-02T05:00:00.000Z',
    },
    {
      priceUsd: '57462.2929728636015241',
      time: 1714629600000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T06:00:00.000Z',
    },
    {
      priceUsd: '57596.3996240546355440',
      time: 1714633200000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T07:00:00.000Z',
    },
    {
      priceUsd: '57669.7948503898871027',
      time: 1714636800000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T08:00:00.000Z',
    },
    {
      priceUsd: '57879.2448705878824941',
      time: 1714640400000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T09:00:00.000Z',
    },
    {
      priceUsd: '57816.8318253461107454',
      time: 1714644000000,
      circulatingSupply: '19692946.0000000000000000',
      date: '2024-05-02T10:00:00.000Z',
    },
    {
      priceUsd: '57821.9173267937048981',
      time: 1714647600000,
      circulatingSupply: '19692946.0000000000000000',
      date: '2024-05-02T11:00:00.000Z',
    },
    {
      priceUsd: '58537.5857861513969580',
      time: 1714651200000,
      circulatingSupply: '19693006.0000000000000000',
      date: '2024-05-02T12:00:00.000Z',
    },
    {
      priceUsd: '58634.5239917690989039',
      time: 1714654800000,
      circulatingSupply: '19693006.0000000000000000',
      date: '2024-05-02T13:00:00.000Z',
    },
    {
      priceUsd: '58479.7446811632706652',
      time: 1714658400000,
      circulatingSupply: '19693006.0000000000000000',
      date: '2024-05-02T14:00:00.000Z',
    },
  ],
  isLoading: false,
}));
