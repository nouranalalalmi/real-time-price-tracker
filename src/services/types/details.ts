export interface AssetDetails {
  id: string;
  rank: string;
  symbol: string;
  name: string;
  supply: string;
  maxSupply: string | null;
  marketCapUsd: string;
  volumeUsd24Hr: string;
  priceUsd: string;
  changePercent24Hr: string;
  vwap24Hr: string;
  explorer: string;
}

export interface AssetHistory {
  priceUsd: string;
  time: number;
  circulatingSupply: string;
  date: string;
}

export interface GetAssetHistoryPayload {
  asset: string;
  range: RangeEnum;
}

export enum RangeEnum {
  DAY = '1d',
  WEEK = '1w',
  MONTH = '1m',
  YEAR = '1y',
}
