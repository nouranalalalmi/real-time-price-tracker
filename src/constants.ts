import { RangeEnum } from './services/types/details';

export const Coins = [
  {
    id: 'bitcoin',
    tickerId: 'btcusdt',
    symbol: 'btc',
    name: 'Bitcoin',
    logo: '/images/btc.png',
  },
  {
    id: 'ethereum',
    tickerId: 'ethusdt',
    symbol: 'eth',
    name: 'Ethereum',
    logo: '/images/eth.png',
  },
  {
    id: 'solana',
    tickerId: 'solusdt',
    symbol: 'sol',
    name: 'Solana',
    logo: '/images/sol.png',
  },
  {
    id: 'cardano',
    tickerId: 'adausdt',
    symbol: 'ada',
    name: 'Cardano',
    ticker: '@ticker',
    logo: '/images/ada.png',
  },
];

export const RangeIntervalMap = {
  [RangeEnum.DAY]: { interval: 'h1', dateRange: 24 },
  [RangeEnum.WEEK]: { interval: 'h6', dateRange: 168 },
  [RangeEnum.MONTH]: { interval: 'h12', dateRange: 210 },
  [RangeEnum.YEAR]: { interval: 'd1', dateRange: 2520 },
};
