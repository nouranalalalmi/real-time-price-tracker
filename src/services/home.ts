import { useQuery } from '@tanstack/react-query';
import axios from 'axios';

import { COINS } from '@/constants';

import { AssetPriceResponse } from './types/home';

export const useGetAssetPrice = () => {
  return useQuery<AssetPriceResponse[], Error>({
    queryKey: ['initPrices'],
    queryFn: async () => {
      const data = await axios.request({
        method: 'GET',
        url: 'https://data-api.binance.vision/api/v3/ticker/24hr',
        params: {
          symbols: JSON.stringify(COINS.map(coin => coin.tickerId.toUpperCase())),
        },
      });

      return data.data;
    },
  });
};
