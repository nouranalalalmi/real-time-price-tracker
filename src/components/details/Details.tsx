import { COINS } from '@/constants';

import { Chart } from './chart/Chart';
import { Overview } from './overview/Overview';

// This function runs at build time and generates the list of all possible paths for dynamic routes
export async function getStaticPaths() {
  const paths = COINS.map(coin => ({
    params: { id: coin.id.toString() },
  }));

  return { paths, fallback: false };
}

// This function runs for each path returned by getStaticPaths and fetches the data for that page
export async function getStaticProps({
  params,
}: {
  params: {
    id: string;
  };
}) {
  const coin = COINS.find(coin => coin.id.toString() === params.id);

  if (!coin) {
    return {
      notFound: true,
    };
  }

  return {
    props: {
      coin,
    },
  };
}

export const Details = () => (
  <>
    <div className="flex w-full flex-col justify-between gap-6 md:flex-row">
      <Chart />
      <div className="w-full md:w-1/4">
        <Overview />
      </div>
    </div>
    <p className="p-2 text-xs font-semibold text-yellow-600">- Data is refetched every 3 minutes</p>
  </>
);
