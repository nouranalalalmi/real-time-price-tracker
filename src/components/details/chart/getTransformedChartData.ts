import dayjs from 'dayjs';
import type { ChartProps } from 'react-chartjs-2';

import { CHART_COLORS } from '@/constants';
import { AssetHistory } from '@/services/types/details';

export const getTransformedChartData = (
  dataHistory: AssetHistory[]
): ChartProps<'line'>['data'] => {
  const labels = dataHistory?.map(item => dayjs(item.date).format('DD-MM-YYYY hh:mm a'));
  const dataPoints = dataHistory?.map(item => Number(item.priceUsd));

  return {
    labels,
    datasets: [
      {
        fill: true,
        label: 'Price',
        data: dataPoints,
        backgroundColor:
          // compare the last data point with the first, if it is is greater than the first, then it is green, else red
          dataPoints && dataPoints.slice(-1)[0] >= dataPoints[0]
            ? CHART_COLORS.TRANSPARENT_GREEN
            : CHART_COLORS.TRANSPARENT_RED,
        segment: {
          borderColor: ctx => {
            // compare each point with the first point, if it is greater than the first, then it is green, else red
            return ctx.p1.parsed.y > dataPoints[0] ? CHART_COLORS.GREEN : CHART_COLORS.RED;
          },
        },
      },
    ],
  };
};
