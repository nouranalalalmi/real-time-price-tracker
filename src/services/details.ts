import { useQuery } from '@tanstack/react-query';
import axios from 'axios';
import dayjs from 'dayjs';

import { RangeIntervalMap } from '@/constants';

import { AssetDetails, AssetHistory, GetAssetHistoryPayload, RangeEnum } from './types/details';

export const useGetAssetDetails = (asset: string) => {
  return useQuery<AssetDetails, Error>({
    queryKey: ['details', asset],
    queryFn: async () => {
      const data = await axios.get(`https://api.coincap.io/v2/assets/${asset}`);
      return data.data.data;
    },
    enabled: !!asset,
    // Refetch the data three minutes
    refetchInterval: 60 * 3000,
  });
};

export const useGetAssetHistory = ({ asset, range = RangeEnum.DAY }: GetAssetHistoryPayload) => {
  return useQuery<AssetHistory[], Error>({
    queryKey: ['history', asset, range],
    queryFn: async () => {
      const data = await axios.request({
        method: 'GET',
        url: `https://api.coincap.io/v2/assets/${asset}/history`,
        params: {
          interval: RangeIntervalMap[range].interval,
          // Subtract the number of days specified by the current selection from the current date
          // and set the time to the start of the day to get the start date
          start: dayjs().subtract(RangeIntervalMap[range].dateRange, 'hours').valueOf(),
          // Set the end date to the end of the current day
          end: dayjs().endOf('day').valueOf(),
        },
      });
      return data.data.data;
    },
    enabled: !!asset,
    refetchOnWindowFocus: false,
    // Refetch the data every three minutes
    refetchInterval: 60 * 3000,
  });
};
