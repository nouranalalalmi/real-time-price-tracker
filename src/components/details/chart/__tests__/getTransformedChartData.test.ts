import { CHART_COLORS } from '@/constants';
import { useGetAssetHistory } from '@/services/details';

import { getTransformedChartData } from '../getTransformedChartData';
jest.mock('@/services/details');

describe('getTransformedChartData', () => {
  const dataHistory = (useGetAssetHistory as jest.Mock)().data || [];
  it('should return the correct chart data', () => {
    const result = getTransformedChartData(dataHistory);

    expect(result).toMatchSnapshot();
  });

  it('should handle empty data history', () => {
    const result = getTransformedChartData([]);

    expect(result.labels).toEqual([]);
    expect(result.datasets[0].data).toEqual([]);
    expect(result.datasets[0].backgroundColor).toEqual(CHART_COLORS.TRANSPARENT_RED);
  });

  it('should return the correct border color', () => {
    const dataHistory = [
      {
        date: '2022-01-01T00:00:00Z',
        priceUsd: '100',
        time: 1641027600,
        circulatingSupply: '18500000',
      },
      {
        date: '2022-01-02T00:00:00Z',
        priceUsd: '200',
        time: 1641114000,
        circulatingSupply: '18500000',
      },
    ];

    const result = getTransformedChartData(dataHistory);

    // Check the border color when the y value is greater than the first data point
    expect(result?.datasets[0]?.segment?.borderColor({ p1: { parsed: { y: 200 } } })).toEqual(
      CHART_COLORS.GREEN
    );

    // Check the border color when the y value is less than the first data point
    expect(result?.datasets[0]?.segment?.borderColor({ p1: { parsed: { y: 50 } } })).toEqual(
      CHART_COLORS.RED
    );

    // Check the border color when the y value is equal to the first data point
    expect(result?.datasets[0]?.segment?.borderColor({ p1: { parsed: { y: 100 } } })).toEqual(
      CHART_COLORS.RED
    );
  });
});
