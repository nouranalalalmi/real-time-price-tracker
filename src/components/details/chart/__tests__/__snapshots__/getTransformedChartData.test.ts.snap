// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTransformedChartData should return the correct chart data 1`] = `
{
  "datasets": [
    {
      "backgroundColor": "rgba(34,197,94, 0.1)",
      "data": [
        57453.46542358884,
        56849.82059461345,
        56989.27449840857,
        57421.71965568798,
        58915.94031893552,
        57038.200775008605,
        57342.452978220164,
        57871.16745201774,
        57904.76527674785,
        58350.587565644215,
        57759.017340192746,
        57173.57867330977,
        57451.83614115858,
        57495.41041375102,
        57362.30026919419,
        57462.2929728636,
        57596.39962405463,
        57669.79485038989,
        57879.244870587885,
        57816.83182534611,
        57821.91732679371,
        58537.585786151394,
        58634.5239917691,
        58479.74468116327,
      ],
      "fill": true,
      "label": "Price",
      "segment": {
        "borderColor": [Function],
      },
    },
  ],
  "labels": [
    "01-05-2024 07:00 pm",
    "01-05-2024 08:00 pm",
    "01-05-2024 09:00 pm",
    "01-05-2024 10:00 pm",
    "01-05-2024 11:00 pm",
    "02-05-2024 12:00 am",
    "02-05-2024 01:00 am",
    "02-05-2024 02:00 am",
    "02-05-2024 03:00 am",
    "02-05-2024 04:00 am",
    "02-05-2024 05:00 am",
    "02-05-2024 06:00 am",
    "02-05-2024 07:00 am",
    "02-05-2024 08:00 am",
    "02-05-2024 09:00 am",
    "02-05-2024 10:00 am",
    "02-05-2024 11:00 am",
    "02-05-2024 12:00 pm",
    "02-05-2024 01:00 pm",
    "02-05-2024 02:00 pm",
    "02-05-2024 03:00 pm",
    "02-05-2024 04:00 pm",
    "02-05-2024 05:00 pm",
    "02-05-2024 06:00 pm",
  ],
}
`;
